holoviews/element/sankey.py

```diff
@@ -0,0 +1,344 @@
+            graph['nodes'].append(node)
+            node_map[index] = node
+
+        links = [element.dimension_values(d) for d in element.dimensions()[:3]]
+        for i, (src, tgt, value) in enumerate(zip(*links)):
+            source, target = node_map[src], node_map[tgt]
+            link = dict(index=i, source=source, target=target, value=value)
+            graph['links'].append(link)
+            source['sourceLinks'].append(link)
+            target['targetLinks'].append(link)
+
+    @classmethod
+    def computeNodeValues(cls, graph):
+        """
+        Compute the value (size) of each node by summing the associated links.
+        """
+        for node in graph['nodes']:
+            source_val = np.sum([l['value'] for l in node['sourceLinks']])
+            target_val = np.sum([l['value'] for l in node['targetLinks']])
+            node['value'] = max([source_val, target_val])
+
+    def computeNodeDepths(self, graph):
+        """
+        Iteratively assign the depth (x-position) for each node.
+        Nodes are assigned the maximum depth of incoming neighbors plus one;
+        nodes with no incoming links are assigned depth zero, while
+        nodes with no outgoing links are assigned the maximum depth.
+        """
+        nodes = graph['nodes']
+        depth = 0
+        while nodes:
+            next_nodes = []
+            for node in nodes:
+                node['depth'] = depth
+                for link in node['sourceLinks']:
+                    if link['target'] not in next_nodes:
+                        next_nodes.append(link['target'])
+            nodes = next_nodes
+            depth +=1
+            if depth > 10000:
+                raise RecursionError('Sankey diagrams only support acyclic graphs.')
+
+        nodes = graph['nodes']
+        depth = 0
+        while nodes:
+            next_nodes = []
+            for node in nodes:
+                node['height'] = depth
+                for link in node['targetLinks']:
+                    if link['source'] not in next_nodes:
+                        next_nodes.append(link['source'])
+            nodes = next_nodes
+            depth +=1
+            if depth > 10000:
+                raise RecursionError('Sankey diagrams only support acyclic graphs.')
+
+        x0, _, x1, _ = self.p.bounds
+        dx = self.p.node_width
+        kx = (x1 - x0 - dx) / (depth - 1)
+        for node in graph['nodes']:
+            d  = node['depth'] if node['sourceLinks'] else depth - 1
+            node['x0'] = x0 + max([0, min([depth-1, np.floor(d)]) * kx])
+            node['x1'] = node['x0'] + dx
+
+    def computeNodeBreadths(self, graph):
+        node_map = OrderedDict()
+        for n in graph['nodes']:
+            if n['x0'] not in node_map:
+                node_map[n['x0']] = []
+            node_map[n['x0']].append(n)
+        
+        _, y0, _, y1 = self.p.bounds
+        py = self.p.node_padding
+
+        def initializeNodeBreadth():
```

Comment:
It might be nice to avoid these kinds of inline function - explicit arguments are nicer than closures...